Payjoin-cli should cache ohttp-keys for re-use

zealsham opened the pull request:
This pr  adds the functionality of ohttp-keys catching to payjoin-cli , ohttp-keys should not be fetched each time and a cached key should be use. Keys expire in 6 months .

## To do 
- [x] it shoudnt still use cached key when the relay-url isnt even present  in config or as command-line flag 
- [x] should handle a case of non-expired key but key being rejected by relay (will be handled in different PR)
- [x]  get rid of unwrap calls that can panic (unhandled)
- [x] should not use the cached keys if --ohttp-keys flag is present  cli command 
- [x] it should be possible to cache multiple ohttp-keys based on relay-url (maybe in subsequent pr ?)

## Pull Request Checklist

Please confirm the following before requesting review:

- [x] I have [disclosed my use of
  AI](https://github.com/payjoin/rust-payjoin/blob/master/.github/CONTRIBUTING.md#ai-assistance-notice)
  in the body of this PR.
- [x] I have read [CONTRIBUTING.md](https://github.com/payjoin/rust-payjoin/blob/master/.github/CONTRIBUTING.md#commits) and **rebased my branch to produce [hygienic commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#committing-patches)**.

zealsham commented:
@arminsabouri the PR is ready

arminsabouri commented on payjoin-cli/src/app/v2/ohttp.rs:
```diff
@@ -77,6 +84,17 @@ async fn fetch_ohttp_keys(
 
+        // try cache for this selected relay first
+        if let Some(cached) = read_cached_ohttp_keys(&selected_relay) {
+            println!("using Cached keys  for relay: {selected_relay}");
```
```suggestion
            println!("using Cached keys for relay: {selected_relay}");
```

nothingmuch reviewed (changes requested):
i think the cache TTL being hard coded to 6 months is a bit problematic. if a directory's keys are lost, this may render its URL useless for that period of time, giving operators no recourse.

instead we should probably use standard HTTP cache control headers, which would allow an operator with high confidence in their deployment to set more aggressive caching parameters.

rfc 9458 says (sections 5.2, 6.4) that a server must generate a response with a 422 status in response to an invalid message, which is what the server sees if it can't decapsulate because the client is using a stale pubkey.

i'm not sure what the implementation and/or testing for that is right now, but we should handle that by fetching the ohttp key and comparing it, erroring if there is an inconsistency but allowing opting in to just overwrite the old key with the new key (i.e. making such errors into warnings)

DanGould commented on payjoin-cli/src/app/v2/ohttp.rs:
```diff
@@ -116,3 +143,49 @@ async fn fetch_ohttp_keys(
+    fetched_at: u64,
+}
+
+fn get_cache_file(relay_url: &url::Url) -> PathBuf {
```
Another note: keys correspond to directories, not relays. Might just be a mistype but it's a concrete distinction I felt necessary to repeat for certainty.

bc1cindy reviewed (commented):
the consumption side looks right 

the `s-maxage` parsing already matches what #1449 will emit

the main piece left is wiring the primitives in (they're defined but not called yet): get_cached_ohttp_keys(directory) → on miss fetch_ohttp_keys_with_cache(...) → store_ohttp_keys(directory, keys, now + ttl)

zealsham committed:
Payjoin-cli should catch ohttp-keys for re-use

This pr  adds the functionality of ohttp-keys
catching to payjoin-cli , ohttp-keys should not
be fetched each time and a cached key should be use when available
This pr complements #1449
